tasks.rb

```diff
@@ -4916,6 +4933,210 @@ def download_epws
+
+      param_type.each_node do |element|
+        next unless element.is_a? Oga::XML::Element
+        next unless (element.name == 'element' || element.name == 'group')
+        next if element.name == 'element' && (element.get('name').nil? && element.get('ref').nil?)
+        next if element.name == 'group' && element.get('ref').nil?
+
+        ancestors = []
+        element.each_ancestor do |node|
+          next if node.get('name').nil?
+          next if node.get('name') == param_type.get('name') # exclude complexType name from element xpath
+
+          ancestors << node.get('name')
+        end
+
+        parent_element_names = ancestors.reverse
+        if element.name == 'element'
+          child_element_name = element.get('name')
+          child_element_name = element.get('ref') if child_element_name.nil? # Backup
+          element_type = element.get('type')
+          element_type = element.get('ref') if element_type.nil? # Backup
+        elsif element.name == 'group'
+          child_element_name = nil # exclude group name from the element's xpath
+          element_type = element.get('ref')
+        end
+        element_xpath = parent_element_names.push(child_element_name)
+        complex_type_or_group_dict[param_type_name][element_xpath] = element_type
+      end
+    end
+  end
+
+  element_xpaths = {}
+  top_level_elements_of_interest = elements_in_sample_files.map { |e| e.split('/')[1].gsub('h:', '') }.uniq
+  top_level_elements_of_interest.each do |element|
+    top_level_element = []
+    top_level_element << element
+    top_level_element_type = element
+    get_element_full_xpaths(element_xpaths, complex_type_or_group_dict, top_level_element, top_level_element_type)
+  end
+
+  # Add enumeration and min/max numeric values
+  rules = {}
+  element_xpaths.each do |element_xpath, element_type|
+    next if element_type.nil?
+
+    # Skip element xpaths not being used in sample files
+    element_xpath_with_prefix = element_xpath.compact.map { |e| "h:#{e}" }
+    context_xpath = element_xpath_with_prefix.join('/').chomp('/')
+    next unless elements_in_sample_files.any? { |item| item.include? context_xpath }
+
+    hpxml_data_type_name = [element_type, '_simple'].join() # FUTURE: This may need to be improved later since enumeration and minimum/maximum values cannot be guaranteed to always be placed within simpleType.
+    hpxml_data_type = hpxml_data_types_dict[hpxml_data_type_name]
+    hpxml_data_type = hpxml_data_types_dict[element_type] if hpxml_data_type.nil? # Backup
+    if hpxml_data_type.nil?
+      fail "Could not find data type name for '#{element_type}'."
+    end
+
+    next if hpxml_data_type[:enums].empty? && hpxml_data_type[:min_inclusive].nil? && hpxml_data_type[:max_inclusive].nil? && hpxml_data_type[:min_exclusive].nil? && hpxml_data_type[:max_exclusive].nil?
+
+    element_name = context_xpath.split('/')[-1]
+    context_xpath = context_xpath.split('/')[0..-2].join('/').chomp('/').prepend('/h:HPXML/')
```

Comment:
Note that I changed from `prepend('//')` to `prepend('/h:HPXML/')`. We should always try to avoid the former since it checks for an HPXML element _anywhere_ in the file, which is much slower than giving it an absolute xpath so that it only looks at the root node.

Reply:
💡